File: backend/app/schemas/waterfall.py
"""Response models for the 3D FFT waterfall (many captures, one channel)."""
from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

SELECTION_MODES = ["last", "oldest", "random"]
STACKING_OLDEST_TO_NEWEST = "oldest_to_newest"


class WaterfallPeakOut(BaseModel):
    """A detected spectral peak within one capture."""
    frequency: float
    amplitude: float


class WaterfallCaptureOut(BaseModel):
    """One FFT spectrum = one upload = one row of the waterfall."""
    capture_number: int = Field(description="1-based position in stacking order (oldest -> newest)")
    upload_id: UUID
    original_filename: Optional[str] = None
    captured_at: datetime
    channel: int
    point_count: int
    frequencies: List[float]
    amplitudes: List[float]
    peaks: List[WaterfallPeakOut] = Field(default_factory=list)


class WaterfallOut(BaseModel):
    sensor_id: UUID
    channel: int
    plot_type: str
    selection_mode: str
    stacking: str = STACKING_OLDEST_TO_NEWEST

    requested_count: int
    returned_count: int
    total_available: int
    skipped_count: int = 0

    sampling_rate_hz: float
    fft_lines: Optional[int] = None
    window: str
    data_type: str

    frequency_min_hz: Optional[float] = None
    frequency_max_hz: Optional[float] = None
    frequency_resolution_hz: Optional[float] = None
    amplitude_min: Optional[float] = None
    amplitude_max: Optional[float] = None

    x_label: str
    y_label: str
    z_label: str

    # ── Appendix A axis contract ────────────────────────────────────────────
    #: This is a STACKED-SPECTRA waterfall: one spectrum per capture file, not
    #: an STFT of a single capture. The two have different Y axes and must not
    #: be conflated — an STFT's Y is elapsed time within one capture, uniform
    #: at hop/Fs, whereas this Y is a list of capture timestamps.
    waterfall_kind: str = "stacked_spectra"
    x_unit: str = "Hz"
    x_scale: str = "linear"
    #: Capture timestamps in stacking order, one per row of `captures`. Plot
    #: against real datetimes or treat as categorical.
    y_values: List[datetime] = Field(default_factory=list)
    y_unit: str = "datetime"
    y_scale: str = "linear"
    #: Captures are selected by recency/randomness, so the gap between adjacent
    #: rows is NOT constant. Never assume a uniform step on this axis.
    y_uniform: bool = False
    z_unit: str
    z_scale: str = "linear"
    #: Linear 0-to-peak, never RMS and never dB.
    amplitude_convention: str = "0-peak"

    sensor_label: Optional[str] = None
    orientation: Optional[str] = None
    mounting_location: Optional[str] = None

    captures: List[WaterfallCaptureOut]

File: frontend/src/types/waterfall.ts
/** Shapes returned by GET /api/v1/measurements/waterfall. */

export const WATERFALL_MODES = ["last", "oldest", "random"] as const;
export type WaterfallMode = (typeof WATERFALL_MODES)[number];

export const WATERFALL_MODE_LABELS: Record<WaterfallMode, string> = {
  last: "Last N",
  oldest: "Oldest N",
  random: "Random",
};

/** Spectra the waterfall can stack — both are already produced by the plot pipeline. */
export const WATERFALL_SPECTRA = ["fft_spectrum", "envelope_spectrum"] as const;
export type WaterfallSpectrum = (typeof WATERFALL_SPECTRA)[number];

export const WATERFALL_SPECTRUM_LABELS: Record<WaterfallSpectrum, string> = {
  fft_spectrum: "FFT Spectrum",
  envelope_spectrum: "Envelope Spectrum",
};

export interface WaterfallPeak {
  frequency: number;
  amplitude: number;
}

export interface WaterfallCapture {
  /** 1-based position in stacking order (oldest -> newest). */
  capture_number: number;
  upload_id: string;
  original_filename: string | null;
  captured_at: string;
  channel: number;
  point_count: number;
  frequencies: number[];
  amplitudes: number[];
  peaks: WaterfallPeak[];
}

export interface WaterfallResponse {
  sensor_id: string;
  channel: number;
  plot_type: string;
  selection_mode: WaterfallMode;
  stacking: string;

  requested_count: number;
  returned_count: number;
  total_available: number;
  skipped_count: number;

  sampling_rate_hz: number;
  fft_lines: number | null;
  window: string;
  data_type: string;

  frequency_min_hz: number | null;
  frequency_max_hz: number | null;
  frequency_resolution_hz: number | null;
  amplitude_min: number | null;
  amplitude_max: number | null;

  x_label: string;
  y_label: string;
  z_label: string;

  sensor_label: string | null;
  orientation: string | null;
  mounting_location: string | null;

  captures: WaterfallCapture[];
}

export interface WaterfallQuery {
  sensorId: string;
  channel: number;
  count: number;
  mode: WaterfallMode;
  plotType: WaterfallSpectrum;
  maxPoints?: number;
  maxPeaks?: number;
  /** Keeps a random selection stable across refetches. */
  seed?: number;
}
